### ISSUE TITLE
dump(config()) in tests throws BindingResolutionException:  Target class [config] does not exist

### ISSUE THREAD

**Issue by scibuff** [laravel/framework#32857]
- Laravel Version: 7.10.3
- PHP Version: 7.3.9-1+ubuntu18.04.1
- Database Driver & Version:
- PHPUnit 8.5.4
- laravel/homestead  (virtualbox, 8.2.1)
- vccw-team/xenial64 (virtualbox, 20180107)
- Vagrant 2.2.6

### Description:

I experience a weird behavior when I try to `dump`  values from configuration files during my tests. Everything works fine when I run tests one by one (one method at a time). But running an entire class (or even just two tests) at once gives me an error 

`Illuminate\Contracts\Container\BindingResolutionException: Target class [config] does not exist.`

### Steps To Reproduce:

Create a simple test, e.g. `php artisan make:test MyBasicTest --unit`

```php
<?php 
namespace Tests\Unit;

use Tests\TestCase;

class MyBasicTest extends TestCase {

    /** @test */
    public function test1(){
        dump( config('app.name') );
        dump( config('app.env') );
        $this->assertTrue(true);
    }

    /** @test */
    public function test2(){
        dump( config('app.name') );
        dump( config('app.env') );
        $this->assertTrue(true);
    }
}
```

Then, running the test as `phpunit --filter test1` or `phpunit --filter test2` is fine, but when I do `phpunit --filter MyBasicTest` I get 

```php
vagrant@homestead:~/code/apps/my-app$ phpunit --filter MyBasicTest
PHPUnit 8.5.4 by Sebastian Bergmann and contributors.

"My App"
"local"
."My App"
E                                                                  2 / 2 (100%)

Time: 854 ms, Memory: 18.00 MB

There was 1 error:

1) Tests\Unit\MyBasicTest::test2
Illuminate\Contracts\Container\BindingResolutionException: Target class [config] does not exist.

/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:807
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:687
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:796
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:633
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:781
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php:119
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php:275
/home/vagrant/code/apps/my-app/tests/Unit/MyBasicTest.php:30

Caused by
ReflectionException: Class config does not exist

/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:805
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:687
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:796
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Container/Container.php:633
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/Application.php:781
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php:119
/home/vagrant/code/apps/my-app/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php:275
/home/vagrant/code/apps/my-app/tests/Unit/MyBasicTest.php:30

ERRORS!
Tests: 2, Assertions: 1, Errors: 1.
```

Basically, the second call to `dump(config('app.env');` inside `test2` throws the exception

p.s. if I had to guess (not knowing the implementation details of the framework) it is as if the `dump` calls in `test1` unbind the config from the app instance and `test2` doesn't manage to finish before that happens

**Comment by taylorotwell** [comment]
Need to use a feature test, not a unit test.

**Comment by scibuff** [comment]
@taylorotwell Please excuse my ignorance, but why would that matter? In any case, I get the same outcome after moving the test to feature

**Comment by ionutantohi** [comment]
@scibuff see this one #30700

**Comment by vmrfriz** [comment]
If you have a setUp method, it should call the parent method.
```php
protected function setUp(): void
{
    parent::setUp();
    // Your code
}
```

If you are trying to use `config()` in dataProvider, you need to call `$this->setUp()` in it:
```php
/**
 * @dataProvider myFirstDataProvider
 */
public function testFromSubdomain($subdomain, $expected)
{
    $email = EmailHelper::fromSubdomain($subdomain);
    $this->assertEquals($expected, $email);
}

public function myFirstDataProvider(): array
{
    $this->setUp();
//  ^^^^^^^^^^^^^^
//     use this

    $default_email = config('mail.recipient');

    return [
        'london' => ['lon', 'london@example.com'],
        'moscow' => ['msk', 'moskva@example.com'],
        'helsinki' => ['helsi', 'helsinki1@example.com'],
        'berlin' => ['berlin', $default_email],
    ];
}
```

The tests are working
```
 PASS  ...\EmailHelperTest
  ✓ from subdomain with data set "london"
  ✓ from subdomain with data set "moscow"
  ✓ from subdomain with data set "helsinki"
  ✓ from subdomain with data set "berlin"
```
